pkg/gpu/cuda/fatbin.go

```diff
@@ -0,0 +1,475 @@
+}
+
+// CubinKernelKey is the key to identify a kernel in a fatbin
+type CubinKernelKey struct {
+	Name      string
+	SmVersion uint32
+}
+
+// CubinKernel holds the information of a CUDA kernel
+type CubinKernel struct {
+	Name        string                          // Name of the kernel
+	attributes  map[nvInfoAttr]nvInfoParsedItem // Attributes of the kernel
+	SymtabIndex int                             // Index of this kernel in the ELF symbol table
+	KernelSize  uint64                          // Size of the kernel in bytes
+	SharedMem   uint64                          // Size of the shared memory used by the kernel
+	ConstantMem uint64                          // Size of the constant memory used by the kernel
+}
+
+// GetKernel returns the kernel with the given name and SM version from the fatbin
+func (fb *Fatbin) GetKernel(name string, smVersion uint32) *CubinKernel {
+	key := CubinKernelKey{Name: name, SmVersion: smVersion}
+	if _, ok := fb.Kernels[key]; !ok {
+		return nil
+	}
+	return fb.Kernels[key]
+}
+
+type fatbinHeader struct {
+	Magic      uint32
+	Version    uint16
+	HeaderSize uint16
+	FatSize    uint64 // not including the header
+}
+
+type fatbinData struct {
+	Kind                    uint16
+	Version                 uint16
+	HeaderSize              uint32
+	PaddedPayloadSize       uint32
+	Unknown0                uint32 // check if it's written into separately
+	PayloadSize             uint32
+	Unknown1                uint32
+	Unknown2                uint32
+	SmVersion               uint32
+	BitWidth                uint32
+	Unknown3                uint32
+	Unknown4                uint64
+	Unknown5                uint64
+	UncompressedPayloadSize uint64
+}
+
+// ParseFatbinFromELFFilePath opens the given path and parses the resulting ELF for CUDA kernels
+func ParseFatbinFromELFFilePath(path string) (*Fatbin, error) {
+	elfFile, err := elf.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open ELF file %s: %w", path, err)
+	}
+	defer elfFile.Close()
+
+	return ParseFatbinFromELFFile(elfFile)
+}
+
+// ParseFatbinFromPath parses the fatbin sections of the given ELF file and returns the information found in it
+func ParseFatbinFromELFFile(elfFile *elf.File) (*Fatbin, error) {
+	fatbin := &Fatbin{
+		Kernels: make(map[CubinKernelKey]*CubinKernel),
+	}
+
+	for _, sect := range elfFile.Sections {
+		// CUDA embeds the fatbin data in sections named .nv_fatbin or __nv_relfatbin
+		if sect.Name != ".nv_fatbin" && sect.Name != "__nv_relfatbin" {
+			continue
+		}
+
+		data, err := sect.Data()
+		if err != nil {
+			return nil, fmt.Errorf("failed to read section %s: %w", sect.Name, err)
+		}
+
+		// The fatbin format will have a header, then a sequence of data headers + payloads.
+		// After the data corresponding to the first header, we might have more headers so
+		// we need to loop until we reach the end of the section data
+		buffer := bytes.NewReader(data)
+		for buffer.Len() > 0 {
```

Comment:
The problem with that is that in case we find an incomplete header (which probably means an error in the parsing code) we wouldn't get an error message.